File: connector/ce/libs/api-clients/java-client/src/main/kotlin/de/sovity/edc/client/EdcClientFactory.java
/*
 * Copyright 2025 sovity GmbH
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 *
 * Contributors:
 *     sovity - init and continued development
 */
package de.sovity.edc.client;

import de.sovity.edc.client.gen.ApiClient;
import de.sovity.edc.client.gen.api.EnterpriseEditionApi;
import de.sovity.edc.client.gen.api.UiApi;
import de.sovity.edc.client.gen.api.UseCaseApi;
import de.sovity.edc.client.oauth2.OAuth2CredentialsAuthenticator;
import de.sovity.edc.client.oauth2.OAuth2CredentialsInterceptor;
import de.sovity.edc.client.oauth2.OAuth2CredentialsStore;
import de.sovity.edc.client.oauth2.OAuth2TokenFetcher;
import lombok.AccessLevel;
import lombok.NoArgsConstructor;
import org.apache.commons.lang3.StringUtils;

/**
 * Builds {@link EdcClient}s.
 */
@NoArgsConstructor(access = AccessLevel.PRIVATE)
public class EdcClientFactory {

    public static EdcClient newClient(EdcClientBuilder builder) {
        var apiClient = new ApiClient()
            .setServerIndex(null)
            .setBasePath(builder.managementApiUrl());

        if (StringUtils.isNotBlank(builder.managementApiKey())) {
            apiClient.addDefaultHeader("X-Api-Key", builder.managementApiKey());
        }

        if (builder.oauth2ClientCredentials() != null) {
            var tokenFetcher = new OAuth2TokenFetcher(builder.oauth2ClientCredentials());
            var handler = new OAuth2CredentialsStore(tokenFetcher);
            var httpClient = apiClient.getHttpClient()
                .newBuilder()
                .addInterceptor(new OAuth2CredentialsInterceptor(handler))
                .authenticator(new OAuth2CredentialsAuthenticator(handler))
                .build();
            apiClient.setHttpClient(httpClient);
        }

        if (builder.customConfigurer() != null) {
            builder.customConfigurer().accept(apiClient);
        }

        return new EdcClient(
            new UiApi(apiClient),
            new UseCaseApi(apiClient),
            new EnterpriseEditionApi(apiClient)
        );
    }
}

File: examples/chat-app/source-starter-for-tutorial/backend/src/main/kotlin/de/sovity/chatapp/services/edc/EdcClientFactory.kt
/*
 * Copyright 2025 sovity GmbH
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 *
 * Contributors:
 *     sovity - init and continued development
 */
package de.sovity.chatapp.services.edc

import de.sovity.edc.client.EdcClient
import io.quarkus.logging.Log
import jakarta.enterprise.context.ApplicationScoped
import jakarta.enterprise.inject.Produces
import org.eclipse.microprofile.config.inject.ConfigProperty
import java.lang.Thread.sleep
import java.time.Duration

@ApplicationScoped
class EdcClientFactory(
    @ConfigProperty(name = "chat-app.edc.management-api.url")
    private val managementApiUrl: String,

    @ConfigProperty(name = "chat-app.edc.management-api.key")
    private val managementApiKey: String
) {

    @Produces
    fun edcClient(): EdcClient {
        val client = EdcClient.builder()
            .managementApiUrl(managementApiUrl)
            .managementApiKey(managementApiKey)
            .build()

        retry(
            times = 60,
            delay = Duration.ofSeconds(1),
            messageOnFailure = "Failed to connect the EDC Management API at $managementApiUrl"
        ) {
            client.testConnection()
            Log.info("Successfully connected to EDC")
        }

        return client
    }

    private fun retry(
        times: Int,
        delay: Duration,
        messageOnFailure: String,
        block: () -> Unit
    ) {
        repeat(times) {
            try {
                block()
                return
            } catch (e: Exception) {
                if (it == times - 1) {
                    throw RuntimeException(messageOnFailure, e)
                }
                try {
                    sleep(delay.toMillis())
                } catch (e: Exception) {
                    throw RuntimeException(e)
                }
            }
        }
    }
}

